### ISSUE TITLE
Properly escape annotation values in ConfigMaps

### ISSUE THREAD

**Issue by fgiorgetti** [skupperproject/skupper#681]
Annotations passed thru `skupper init --annotations` containing `=` or `,` in their values, should be properly escaped.

**Comment by waldner** [comment]
Any news on this one? I'm trying to use an annotation whose value contains commas in the `skupper-site` confgmap, and it looks like it's completely ignored:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: skupper-site
data:
  name: "mainsite"
  cluster-local: "false"
  console: "true"
  console-authentication: internal
  console-user: "barney"
  console-password: "rubble"
  edge: "false"
  flow-collector: "true"
  ingress: "nginx-ingress-v1"
  ...
  router-service-annotations: "foo.bar.com/exclude=item1,item2,item3"   # ignored
```

Since I'm using ArgoCD, I also saw https://github.com/skupperproject/skupper/issues/1122 and tried to use serverside apply, no difference.

**Comment by waldner** [comment]
Ok, my bad, I was expecting the annotations to be applied to the `Pod`, not to the `Service` (although its name is `router-service-annotations`, so that is a clue isn't it). They are indeed applied to the `Service`, but I need to apply them to the pod. I've seen https://github.com/skupperproject/skupper/pull/444 which annotates all pods, is there a way to only annotate a deployment? Going by https://github.com/skupperproject/skupper/issues/930 it looks like there's no way, is that still the case?

**Comment by grs** [comment]
Yes, unfortunately there is still no way to label/annotate the deployment itself rather than the pods. This is something we are currently looking at, trying to come up with a more comprehensive and coherent solution to the customisation of skupper created (or augmented) resources.

**Comment by waldner** [comment]
Thanks, As a first step, would you be willing to consider a PR adding `{router,controller}-pod-annotations` or similar to `skupper-site` and command line? I'm not saying I'll be able to do it, but I can try at least, if there is interest.

**Comment by grs** [comment]
Would it work for you if the annotations on skupper-site were applied to the deployment as well as to the pods? (Currently they are just applied to the pods created from the deployment). I.e. is it necessary to *not* put those annotations in the podTemplateSpec? That would be the simplest fix.

**Comment by waldner** [comment]
My main concern is not whether annotations are applied to pods or deployments (although I need them on the pods), but rather to be able to selectively apply them only to the router or the controller pods.

**Comment by grs** [comment]
Ok, sorry, I understand now. Yes, we would accept a PR with options for differentiating router v. service controller pods.

**Comment by waldner** [comment]
Are there instructions for development? I'm having trouble because, even without changing anything, a `site-controller` image built from git gives me a nonworking service-controller. Here's what I'm doing:

```
$ scripts/libpod-generate.sh
$ docker build -t myregistry.com/site-controller-test -f Dockerfile.site-controller .
# use the site-controller-test image to create a deployment as follows:

apiVersion: v1
kind: ServiceAccount
metadata:
  name: skupper-site-controller
  labels:
    application: skupper-site-controller
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-site-controller
  namespace: myns
spec:
  replicas: 1
  selector:
    matchLabels:
      application: skupper-site-controller
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/part-of: skupper
        application: skupper-site-controller
    spec:
      containers:
        - name: site-controller
          image: >-
            myregistry.com/site-controller-test
          env:
            - name: WATCH_NAMESPACE
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.namespace
            - name: QDROUTERD_IMAGE
              value: >-
                quay.io/skupper/skupper-router:2.4.1-2
            - name: SKUPPER_SERVICE_CONTROLLER_IMAGE
              value: >-
                quay.io/skupper/service-controller:1.4.1
            - name: SKUPPER_CONFIG_SYNC_IMAGE
              value: >-
                quay.io/skupper/config-sync:1.4.1
            - name: SKUPPER_FLOW_COLLECTOR_IMAGE
              value: >-
                quay.io/skupperflow-collector:1.4.1
          imagePullPolicy: IfNotPresent
          securityContext:
            capabilities:
              drop:
                - ALL
            runAsNonRoot: true
            allowPrivilegeEscalation: false
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      serviceAccountName: skupper-site-controller
      serviceAccount: skupper-site-controller
      securityContext:
        runAsUser: 65534
        runAsNonRoot: true
        seccompProfile:
          type: RuntimeDefault
      schedulerName: default-scheduler
  strategy:
    type: Recreate
  revisionHistoryLimit: 10
  progressDeadlineSeconds: 600
```
Once I apply this `skupper-site`:

```
data:
  name: "mysite"
  cluster-local: "false"
  console: "true"
  console-authentication: internal
  console-user: "barney"
  console-password: "rubble"
  edge: "false"
  flow-collector: "true"
  ingress: "nginx-ingress-v1"
  ingress-host: example.com
  router-ingress-host: example.com
  controller-ingress-host: example.com
  router-console: "true"
  service-controller: "true"
  service-sync: "true"
```
what happens is that the router and service-controller pods are created, but `service-controller` doesn't boot with the following error:

```
2023/08/05 15:17:20 open /etc/service-controller/certs/tls.crt: no such file or directory
```
because the secret `skupper-claims-server` is not mounted into the pod (it's not even created, for that matter).
However, replacing my custom built image with `quay.io/skupper/site-controller:1.4.1` works just fine, but then again, I'm not sure I'm doing things correctly, so before starting to dig down the rabbit hole, I'd like to confirm that I'm setting this up in the right way.

Thanks

**Comment by grs** [comment]
Is your site-controller built from the 1.4 branch? There is a change on main now where the claims have been moved into config-sync. It sounds like perhaps you have the site-controller version from main but are trying to use that with mages from 1.4?

**Comment by waldner** [comment]
Yes, as I said I was using site-controller from `main` and the rest from 1.4.1, so it was not working. Building everything out of `main` made it work.

**Comment by fgiorgetti** [comment]
Closing as resolved through #1205
